lib/rules/callback-return.js

```diff
@@ -0,0 +1,136 @@
+     * @returns {ASTNode} The matched node or undefined.
+     */
+    function findClosestParentOfType(node, types) {
+        if (!node.parent) {
+            return null;
+        }
+        if (types.indexOf(node.parent.type) === -1) {
+            return findClosestParentOfType(node.parent, types);
+        }
+        return node.parent;
+    }
+
+    /**
+     * Check to see if a CallExpression is in our callback list.
+     * @param {ASTNode} node The node to check against our callback names list.
+     * @returns {Boolean} Whether or not this function matches our callback name.
+     */
+    function isCallback(node) {
+        return node.callee.type === "Identifier" && callbacks.indexOf(node.callee.name) > -1;
+    }
+
+    /**
+     * Determines whether or not the callback is part of a callback expression.
+     * @param {ASTNode} node The callback node
+     * @param {ASTNode} parentNode The expression node
+     * @returns {boolean} Whether or not this is part of a callback expression
+     */
+    function isCallbackExpression(node, parentNode) {
+
+        // ensure the parent node exists and is an expression
+        if (!parentNode || parentNode.type !== "ExpressionStatement") {
+            return false;
+        }
+
+        // cb()
+        if (parentNode.expression === node) {
+            return true;
+        }
+
+        // special case for cb && cb() and similar
+        if (parentNode.expression.type === "BinaryExpression" || parentNode.expression.type === "LogicalExpression") {
+            if (parentNode.expression.right === node) {
+                return true;
+            }
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    return {
+        "CallExpression": function(node) {
+
+            // if we"re not a callback we can return
+            if (!isCallback(node)) {
+                return;
+            }
+
+            // find the closest block, return or if statement
+            var closestBlock = findClosestParentOfType(node, ["BlockStatement", "ReturnStatement", "IfStatement"]),
+                lastItem;
+
+            // if there wasn't one of those nearby we're probably fine
+            if (!closestBlock) {
+                return;
+            }
+
+            // if our parent is a return we know we're ok
+            if (closestBlock.type === "ReturnStatement") {
+                return;
+            }
+
+            // block statements are part of functions and most if statements
+            if (closestBlock.type === "BlockStatement") {
+
+                // find the last item in the block
+                lastItem = closestBlock.body[closestBlock.body.length - 1];
+
+                // if the callback is the last thing in a block that might be ok
+                if (isCallbackExpression(node, lastItem)) {
```

Comment:
I think most of the real-world cases we were hoping to catch were inside of `IfStatements`. I'm willing to remove that though and just say you *always* have to return with a callback everywhere. Thoughts?